### ISSUE TITLE
Regression: Error on indexing across compatible type parameters if constraint contains `undefined`

### ISSUE THREAD

**Issue by dragomirtitian** [microsoft/TypeScript#44166]
# Bug Report

### 🔎 Search Terms

assignability regression 

### 🕗 Version & Regression Information

- This changed between versions 4.2.3 and 4.3.0-beta

### ⏯ Playground Link

[Playground Link](https://www.typescriptlang.org/play?ts=4.3.0-beta#code/MYGwhgzhAEAqYCMQFMBClkFEB2BXAtgDwBQ0Z0ACrkgJbADKADssNMgB4Auy2AJjACUWAewBOvQgGtkAT2EAzaAEls3UdjAgmLADTQw2GQD4dpcirUatzVh259BI8VNkLK1EHW3A9B42QAfaFw+ZHkabGReaABeYNDwyN4jaABvMzJ8AAoASjSM8jIUTmgab1joPBAQAEJ9GAtkdU1vAG5iAsKAei7oTFFRMVLsaAAWADoAJnq2dhtuXnHoADlhNgGhiLHxgGZoQBlyQsLO8jKbAG0q2pnpOUVG5usWAF0Oo-Ievo3RYe3phBYYFwEGQpXKNBgjGEUBoSBk8V4YQiUWggFByE5kT79QY-LYTPYA4BAkHQW5uKi0Bg2ADk0CJ2GwwhKAOCIOinDWEUR7GgnBkzGUqiaVnK+wxYIuVzqkFJrkUFM8VJe4vFNEUWTOLFiMTiIURiSieXS73eomQnFw6naJugAF9xZ8APKSX4TaYGaL4mYceYo9E2zXAS64arSmBk+5Cx7eV4O3rO11TPRNHGJvYHVXeYOhm5y9yUmPQFWFe32oA)

### 💻 Code

```ts
class TableBaseEnum<
    PublicSpec extends Record<keyof InternalSpec, any>,
    InternalSpec extends Record<keyof PublicSpec, any>  | undefined = undefined> {
    m() {
        let iSpec = null! as InternalSpec;

        // Error in 4.2 as expected. No error in 4.3 ❌        
        iSpec[null! as keyof InternalSpec]

        // Error in 4.2 because iSpec is possibly undefined ✅
        // Error in 4.3 because keyof PublicSpec' cannot be used to index type InternalSpec ❌
        iSpec[null! as keyof PublicSpec]
        
        if (iSpec === undefined) {
            return;
        }
        // Ok in 4.2 and 4.3 as expected ✅
        iSpec[null! as keyof InternalSpec]

        // Ok in 4.2, error in 4.3  ❌❗
        iSpec[null! as keyof PublicSpec] 
        
    }
}
```

### 🙁 Actual behavior

- before the check `iSpec` is indexable with `keyof InternalSpec` which seems like a safe null checks issue
- after `iSpec` is checked, you can index with `keyof InternalSpec` but not `keyof PublicSpec` (even though PublicSpec is `Record<keyof InternalSpec, any>`)

### 🙂 Expected behavior

The behavior in 4.2 seems correct. If `iSpec` is not checked index access should give a `possibly undefined` error (it does not in 4.3) and after the check `iSpec` should be indexable with `keyof InternalSpec` or `keyof PublicSpec`
